controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/deployment/DeploymentApiHandler.java

```diff
@@ -132,4 +150,41 @@ private static String toString(DeploymentSpec.UpgradePolicy upgradePolicy) {
         return upgradePolicy.name();
     }
 
+    // ----------------------------- Utilities to pick out the relevant JobStatus -- filter chains should mirror the ones in VersionStatus
+
+    /** The first upgrade job to fail on this version, for this application */
+    private Optional<JobStatus> firstFailingOn(Version version, Application application) {
+        return JobList.from(application)
+                .failing()
+                .not().failingApplicationChange()
+                .not().failingBecause(outOfCapacity)
+                .lastCompleted().on(version)
+                .asList().stream()
+                .min(comparing(job -> job.lastCompleted().get().at()));
+    }
+
+    /** The number of production jobs for this application */
+    private int productionJobsFor(Application application) {
+        return JobList.from(application)
+                .production()
+                .size();
+    }
+
+    /** The number of production jobs with last success on the given version, for this application */
+    private int productionSuccessesFor(Version version, Application application) {
+        return JobList.from(application)
+                .production()
+                .lastSuccess().on(version)
+                .size();
+    }
+
+    /** The last triggered upgrade to this version, for this application */
+    private Optional<JobStatus> lastDeployingTo(Version version, Application application) {
+        return JobList.from(application)
+                .running(controller.applications().deploymentTrigger().jobTimeoutLimit())
+                .lastTriggered().upgrade()
+                .asList().stream()
+                .max(comparing(job -> job.lastTriggered().get().at()));
+    }
+
```

Comment:
I'll try to merge as much as possible of this with what it should mirror, from VersionStatus, and put it in JobList. ... In a new PR.